dascore/proc/whitening.py

```diff
@@ -0,0 +1,161 @@
+    freq_smooth_size
+        Size (in Hz) of moving average window, used to compute the spectrum
+        before whitening
+
+    Notes
+    -----
+    1) The FFT result is divided by the smoothed spectrum before inverting
+       back to time-domain signal. The phase is not changed.
+
+    2) Amplitude is NOT preserved
+
+    3) The output is also bandpass filtered to the desired frequency range,
+    but the smoothing window includes frequencies outside that range
+
+    Example
+    -------
+
+    import dascore as dc
+    import numpy as np
+    import matplotlib.pyplot as plt
+    import scipy.signal as spec
+    from dascore.units import Hz
+    import numpy.fft as fft
+
+    def plot_spectrum(x,T,ax):
+        fftsig=np.abs(fft.fft(x))
+        fftlen=fftsig.size
+        fftsig=fftsig[0:int(fftlen/2)+1]
+        freqvec=np.linspace(0,0.5/T,fftsig.size)
+        ax.plot(freqvec, fftsig)
+        ax.set_xlabel('frequency [Hz]')
+        ax.set_ylabel('Amplitude (|H(w)|)')
+
+    patch = (
+        dc.get_example_patch('dispersion_event')
+    )
+    patch = patch.resample(time=(200 * Hz))
+
+    filt_patch = patch.pass_filter(time=(5, 60))
+    white_patch = filt_patch.whitening(freq_range=[10,50],freq_smooth_size=10)
+
+    fig,((ax1,ax2),(ax3,ax4)) = plt.subplots(2,2,figsize=(10,7))
+    ax1.plot(filt_patch.data[50,:])
+    ax1.set_title('Original data, distance = 50 m')
+    ax2.plot(white_patch.data[50,:])
+    ax2.set_title('Whitened data, distance = 50 m')
+
+    plot_spectrum(filt_patch.data[50,:],1/200,ax3)
+    ax3.set_title('Original data, distance = 50 m')
+    plot_spectrum(white_patch.data[50,:],1/200,ax4)
+    ax4.set_title('Whitened data, distance = 50 m')
+    plt.show()
+    """
+    patch_cop = patch.convert_units(distance="m").transpose("distance", "time")
+    time = patch_cop.coords.get_array("time")
+    dt = (time[1] - time[0]) / np.timedelta64(1, "s")
+
+    if freq_range is None:
+        freq_range = [0, 1 / 2 / dt]
+    else:
+        if not np.size(freq_range) == 2:
+            msg = "Frequency range must include two values"
+            raise ParameterError(msg)
+
+        if freq_range[0] < 0 or freq_range[1] < 0:
+            msg = "Minimal and maximal frequencies have to be non-negative"
+            raise ParameterError(msg)
+
+        if freq_range[0] >= freq_range[1]:
+            msg = "Frequency range must be increasing"
+            raise ParameterError(msg)
+
+    if freq_smooth_size is None:
+        freq_smooth_size = freq_range[1] - freq_range[0]
+    else:
+        if freq_smooth_size <= 0:
+            msg = "Frequency smoothing size must be positive"
+            raise ParameterError(msg)
+
+    # dt = (time[1] - time[0]) / np.timedelta64(1, "s")
+    (nchan, nt) = patch_cop.data.shape
+    nf = 2 ** (math.ceil(math.log(nt, 2)))
+    h_nf = int(nf / 2)
+    fft_d = np.zeros([nchan, nf], dtype=complex)
+    whitened_data = np.zeros([nchan, nt])
```

Comment:
I think a lot of this is duplicated from [these lines](https://github.com/DASDAE/dascore/blob/46281136692e2423d625d439c41675046e1da97a/dascore/proc/filter.py#L155-L158) in `pass_filter`. Do you mind seeing if you can just reuse that logic (which is already tested and supports units)?

Reply:
Done. I all removed unnecessary testing units.